crux_core/src/command/context.rs

```diff
@@ -215,7 +216,14 @@ impl<T: Unpin + Send> Future for ShellRequest<T> {
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         match self.inner.poll_unpin(cx) {
             Poll::Ready((Some(output), _rest)) => Poll::Ready(output),
-            Poll::Ready((None, _rest)) => Poll::Pending,
+            Poll::Ready((None, rest)) => {
+                // We need to make sure that the inner future can be polled again,
+                // because we could be used in a select! and polled repeatedly, even though
+                // the request on the shell side is a goner.
+                self.inner = rest.into_future();
+
+                Poll::Pending
+            }
```

Comment:
nit: I think this would be clearer by either Fusing the inner future

```
pub struct ShellRequest<T: Unpin + Send> {
    inner: Fuse<StreamFuture<ShellStream<T>>>,
}
```

Or by using the `is_terminated` on `self.inner` to properly show that if we get a `Poll::Ready((None, _rest))` the future will never resolved and stay pending


```
fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
    if self.inner.is_terminated() {
        Poll::Pending
    } else {
        ...
    }
}
```

Reply:
Fair point! Can't quite decide which one I like better 🤔

Reply:
Decided to go with `Fuse<T>`